### ISSUE TITLE
Prometheus "out-of-order timestamp" error due to metrics relabeling conflict

### ISSUE THREAD

**Issue by kien-truong** [k8ssandra/k8ssandra#969]
## Bug Report

**Describe the bug**
K8ssandra is currently exporting 2 metrics `collectd_mcac_histogram_count_total` and `collectd_mcac_meter_count_total`. However, using the default relabeling rules, these metrics are being relabeled to the same metrics. This will confuse Prometheus and can cause `out-of-order` timestamp error if the metrics have different timestamps. 

For example:

```
collectd_mcac_histogram_count_total{mcac="org.apache.cassandra.metrics.table.write_latency.system_schema.types",instance="10.11.66.20",mcac_filtered="true",cluster="cluster",dc="dc1",rack="rack1"} 2 1626265466529
collectd_mcac_meter_count_total{mcac="org.apache.cassandra.metrics.table.write_latency.system_schema.types",instance="10.11.66.20",mcac_filtered="true",cluster="cluster",dc="dc1",rack="rack1"} 2 1626265466528
```

After relabelings

```
mcac_table_write_latency{mcac="org.apache.cassandra.metrics.table.write_latency.system_schema.types",instance="10.11.66.20",mcac_filtered="true",cluster="cluster",dc="dc1",rack="rack1", table="types", keyspace="system_schema"} 2 1626265466529
mcac_table_write_latency{mcac="org.apache.cassandra.metrics.table.write_latency.system_schema.types",instance="10.11.66.20",mcac_filtered="true",cluster="cluster",dc="dc1",rack="rack1", table="types", keyspace="system_schema"} 2 1626265466528
```

Because `1626265466528` is before `1626265466529` an `out-of-order timestamp error` will be reported.

**To Reproduce**
Deploy k8ssandra with Prometheus integration using default configuration.

**Expected behavior**
No relabeling conflict and no error reported by Prometheus.

**Environment (please complete the following information):**

* Helm charts version info

```
NAME            NAMESPACE       REVISION        UPDATED                                 STATUS          CHART           APP VERSION
cassandra       cassandra       16              2021-07-07 16:14:03.704604036 +0700 +07 deployed        k8ssandra-1.2.0            
```

* Kubernetes version information:

```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.9", GitCommit:"9dd794e454ac32d97cde41ae10be801ae98f75df", GitTreeState:"clean", BuildDate:"2021-03-18T01:09:28Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.9", GitCommit:"9dd794e454ac32d97cde41ae10be801ae98f75df", GitTreeState:"clean", BuildDate:"2021-03-18T01:00:06Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
```

* Kubernetes cluster kind: `Baremetal`

┆Issue is synchronized with this [Jira Bug](https://k8ssandra.atlassian.net/browse/K8SSAND-712) by [Unito](https://www.unito.io)
┆Reviewer: 5b3c004b70b3b52e40050e56
┆Affected Versions: k8ssandra-1.2.0
┆Fix Versions: k8ssandra-1.4.0
┆Issue Number: K8SSAND-712
┆Priority: Medium

**Comment by tah-mas** [comment]
We are seeing this issue as well. Out of order timestamp reported as soon as the install completes, and after awhile the metrics timestamp starts lagging behind (https://github.com/datastax/metric-collector-for-apache-cassandra/issues/43). Can we please have an ETA for a fix.

**Comment by tah-mas** [comment]
Here you go. 
helm3 install k8ssandra k8ssandra/k8ssandra --values eks.cassandra.values.yaml

-- eks.cassandra.values.yaml
<pre>
<code>
cassandra:
  clusterName: dev-cluster
  # Version of Apache Cassandra to deploy
  version: "4.0.0"

  # Configuration for the /var/lib/cassandra mount point
  cassandraLibDirVolume:
    storageClass: gp2
    size: 50Gi

  heap:
   size: 6G

  resources:
    requests:
      cpu: 2000m
      memory: 12Gi
    limits:
      cpu: 2000m
      memory: 12Gi

  # This key defines the logical topology of your cluster. The rack names and
  # labels should be updated to reflect the Availability Zones where your EKS
  # cluster is deployed.
  datacenters:
  - name: dev-dc
    size: 3
    racks:
    - name: eu-west-2a
      affinityLabels:
        topology.kubernetes.io/zone: eu-west-2a
    - name: eu-west-2b
      affinityLabels:
        topology.kubernetes.io/zone: eu-west-2b
    - name: eu-west-2c
      affinityLabels:
        topology.kubernetes.io/zone: eu-west-2c

  ingress:
    enabled: true

reaper:
  enabled: true
  autoschedule: true
  autoschedule_properties:
    initialDelayPeriod: PT15S
    periodBetweenPolls: PT10M
    timeBeforeFirstSchedule: PT5M
    scheduleSpreadPeriod: PT12H

stargate:
  enabled: false

medusa:
  enabled: true
  storage: s3

  # Reference the Terraform output for the correct bucket name to use here.
  bucketName: XXXXX

  # The secret here must align with the value used in the previous section.
  storageSecret: medusa-bucket-key

  storage_properties:
    region: eu-west-2
    secure: true

  podStorage:
    storageClass: gp2
    size: 50Gi


kube-prometheus-stack:
  # -- Controls whether the kube-prometheus-stack chart is used at all.
  # Disabling this parameter prevents all monitoring components from being
  # installed.
  enabled: true
  coreDns:
    enabled: true
  kubeApiServer:
    enabled: true
  kubeControllerManager:
    enabled: true
  kubeDns:
    enabled: true
  kubeEtcd:
    enabled: true
  kubeProxy:
    enabled: true
  kubeScheduler:
    enabled: true
  kubeStateMetrics:
    enabled: true
  kubelet:
    enabled: true
  nodeExporter:
    enabled: true

  alertmanager:
    # Disabled for now while we build out a set of default alerts
    enabled: true
    serviceMonitor:
      selfMonitor: false

grafana:
  # -- Provisions an instance of Grafana and wires it up with a DataSource
  # referencing this Prometheus installation
  enabled: true
</code>
</pre>

**Comment by adejanovski** [comment]
I'm able to reproduce the error when deploying K8ssandra locally, without any exotic setting. Setting Prometheus logging level to `debug` shows precisely which metrics are being problematic:

```
level=debug ts=2021-09-06T15:08:25.636Z caller=scrape.go:1411 component="scrape manager" scrape_pool=k8ssandra/k8ssandra-prometheus-k8ssandra/0 target=http://10.244.1.38:9103/metrics msg="Out of order sample" series="collectd_mcac_meter_count_total{mcac=\"org.apache.cassandra.metrics.keyspace.validation_time.system_distributed\",instance=\"10.244.1.38\",mcac_filtered=\"true\",cluster=\"mycluster\",dc=\"dc1\",rack=\"default\"}"
level=debug ts=2021-09-06T15:08:25.636Z caller=scrape.go:1411 component="scrape manager" scrape_pool=k8ssandra/k8ssandra-prometheus-k8ssandra/0 target=http://10.244.1.38:9103/metrics msg="Out of order sample" series="collectd_mcac_meter_count_total{mcac=\"org.apache.cassandra.metrics.keyspace.view_lock_acquire_time.system\",instance=\"10.244.1.38\",mcac_filtered=\"true\",cluster=\"mycluster\",dc=\"dc1\",rack=\"default\"}"
level=debug ts=2021-09-06T15:08:25.636Z caller=scrape.go:1411 component="scrape manager" scrape_pool=k8ssandra/k8ssandra-prometheus-k8ssandra/0 target=http://10.244.1.38:9103/metrics msg="Out of order sample" series="collectd_mcac_meter_count_total{mcac=\"org.apache.cassandra.metrics.keyspace.view_lock_acquire_time.system_distributed\",instance=\"10.244.1.38\",mcac_filtered=\"true\",cluster=\"mycluster\",dc=\"dc1\",rack=\"default\"}"
level=debug ts=2021-09-06T15:08:25.636Z caller=scrape.go:1411 component="scrape manager" scrape_pool=k8ssandra/k8ssandra-prometheus-k8ssandra/0 target=http://10.244.1.38:9103/metrics msg="Out of order sample" series="collectd_mcac_meter_count_total{mcac=\"org.apache.cassandra.metrics.keyspace.view_read_time.system_traces\",instance=\"10.244.1.38\",mcac_filtered=\"true\",cluster=\"mycluster\",dc=\"dc1\",rack=\"default\"}"
level=debug ts=2021-09-06T15:08:25.636Z caller=scrape.go:1411 component="scrape manager" scrape_pool=k8ssandra/k8ssandra-prometheus-k8ssandra/0 target=http://10.244.1.38:9103/metrics msg="Out of order sample" series="collectd_mcac_meter_count_total{mcac=\"org.apache.cassandra.metrics.keyspace.write_latency.system_distributed\",instance=\"10.244.1.38\",mcac_filtered=\"true\",cluster=\"mycluster\",dc=\"dc1\",rack=\"default\"}"
level=warn ts=2021-09-06T15:08:25.844Z caller=scrape.go:1369 component="scrape manager" scrape_pool=k8ssandra/k8ssandra-prometheus-k8ssandra/0 target=http://10.244.1.38:9103/metrics msg="Error on ingesting out-of-order samples" num_dropped=136
```

I've narrowed down the issue to MCAC generating both a histogram and a meter, both containing a `count` metric when outputting Timers: https://github.com/datastax/metric-collector-for-apache-cassandra/blob/master/src/main/java/com/datastax/mcac/UnixSocketClient.java#L757-L758

This generates a `collectd_mcac_meter_count_total` and `collectd_mcac_histogram_count_total` that will get the exact same labels in K8ssandra, which Prometheus doesn't seem to appreciate (my skills in the are are fairly low, but it seems like the metric name isn't enough to have Prometheus distinguish them...).

I'm able to fix the issue by adding a new relabeling rule to add a `kind` label which will get either `histogram` or `meter` as a value, making Prometheus happy.

I'll push a PR shortly with the fix.

**Comment by adejanovski** [comment]
PR is available with the fix: #1086

**Comment by tah-mas** [comment]
Thanks @adejanovski, I have manually applied the patch for now and that seems to fix the issue. Now I am back to "error ingesting samples that are too old or are too far in the future". (https://github.com/datastax/metric-collector-for-apache-cassandra/issues/43)
